File: 2024/04/cpp/04.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long      ll;
typedef vector<int>    vi;
typedef vector<bool>   vb;
typedef pair<int, int> ii;

const int INF = 0x3f3f3f3f;
const ll  INFLL = 0x3f3f3f3f3f3f3f3fll;

#define each(x, s)  for (auto& x : s)
#define loop(x)     for (int i = 0;i < x;i++)
#define vloop(v, x) for (int v = 0;v < x;v++)
#define iter(a)     a.begin(), a.end()
#define riter(a)    a.rbegin(), a.rend()
#define endl        "\n"

int task1(vector<string> input) {
  int n = input.size();
  int m = input[0].size();
  string xmas = "XMAS";
  vector<ii> dirs;
  loop(3) vloop(j, 3) if (i != 1 or j != 1) dirs.emplace_back(i - 1, j - 1);
  auto check = [&] (int x, int y, int dx, int dy) {
    loop(4) {
      int nx = x + i * dx;
      int ny = y + i * dy;
      if (nx < 0 or nx >= n or ny < 0 or ny >= m or input[nx][ny] != xmas[i]) return false;
    }
    return true;
  };
  int ans = 0;
  loop(n) vloop(j, m) {
    if (input[i][j] != 'X') continue;
    for (auto [dx, dy] : dirs) {
      if (check(i, j, dx, dy)) ans++;
    }
  }
  return ans;
}

int task2(vector<string> input) {
  int n = input.size();
  int m = input[0].size();
  auto check = [&] (int x, int y) {
    string d1, d2;
    d1 += input[x - 1][y - 1];
    d1 += input[x + 1][y + 1];
    d2 += input[x - 1][y + 1];
    d2 += input[x + 1][y - 1];
    sort(iter(d1));
    sort(iter(d2));
    if (d1 == "MS" and d2 == "MS") return true;
    return false;
  };
  int ans = 0;
  for (int i = 1;i < n -1;i++) for (int j = 1;j < m - 1;j++) {
    if (input[i][j] != 'A') continue;
    if (check(i, j)) ans++;
  }
  return ans;
}

int main() {
  ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
  
  vector<string> input;
  string s;
  while (getline(cin, s)) input.push_back(s);
  cout << "Task 1: " << task1(input) << endl;
  cout << "Task 2: " << task2(input) << endl;
  return 0;
}

File: 2024/04/rust/04.rs
use std::io;

fn check(input: &Vec<String>, x: i32, y: i32, dx: i32, dy: i32) -> i32 {
    let xmas = "XMAS";
    let mut at = 0;
    let n = input.len() as i32;
    let m = input[0].len() as i32;
    let mut ans = 0;
    // start at input[x][y] and sum iterate using dx, dy
    let mut i = x;
    let mut j = y;
    while i >= 0 && i < n && j >= 0 && j < m {
        let c = input[i as usize].chars().nth(j as usize).unwrap();
        let c2 = xmas.chars().nth(at as usize).unwrap();
        if c == c2 {
            at += 1;
            if at == xmas.len() as i32 {
                ans += 1;
                at = 0;
            }
        } else if c == 'X' {
            at = 1;
        } else {
            at = 0;
        }
        i += dx;
        j += dy;
    }
    ans
}

fn task1(input: Vec<String>) -> i32 {
    let mut ans = 0;
    let n = input.len() as i32;
    let m = input[0].len() as i32;

    for i in 0..n { ans += check(&input, i, 0, 0, 1); }
    for i in 0..n { ans += check(&input, i, m - 1, 0, -1); }
    for i in 0..m { ans += check(&input, 0, i, 1, 0); }
    for i in 0..m { ans += check(&input, n - 1, i, -1, 0); }
    for i in 0..n { ans += check(&input, i, 0, 1, 1); }
    for i in 1..m { ans += check(&input, 0, i, 1, 1); }
    for i in 0..n { ans += check(&input, i, m - 1, -1, -1); }
    for i in 1..m { ans += check(&input, n - 1, m - 1 - i, -1, -1); }
    for i in 0..n { ans += check(&input, i, m - 1, 1, -1); }
    for i in 1..m { ans += check(&input, 0, m - i - 1, 1, -1); }
    for i in 0..n { ans += check(&input, i, 0, -1, 1); }
    for i in 1..m { ans += check(&input, n - 1, i, -1, 1); }
    ans
}

fn sort(s: String) -> String {
    let mut v: Vec<char> = s.chars().collect();
    v.sort();
    v.into_iter().collect()
}

fn task2(input: Vec<String>) -> i32 {
    let mut ans = 0;
    let n = input.len() as i32;
    let m = input[0].len() as i32;
    for i in 1..n-1 {
        for j in 1..m-1 {
            if input[i as usize].chars().nth(j as usize).unwrap() != 'A' { continue; }
            // get adjacent chars in NW and SE directions
            let mut s1 = String::new();
            s1.push(input[(i - 1) as usize].chars().nth((j - 1) as usize).unwrap());
            s1.push(input[(i + 1) as usize].chars().nth((j + 1) as usize).unwrap());
            let mut s2 = String::new();
            s2.push(input[(i - 1) as usize].chars().nth((j + 1) as usize).unwrap());
            s2.push(input[(i + 1) as usize].chars().nth((j - 1) as usize).unwrap());
            s1 = sort(s1);
            s2 = sort(s2);
            if s1.eq("MS") && s2.eq("MS") {
                ans += 1;
            }
        }
    }
    ans
}

fn main() {
    let mut line = String::new();
    let mut input = Vec::new();
    while io::stdin().read_line(&mut line).unwrap() > 0 {
        input.push(line.clone());
        line.clear();
    }
    println!("Task 1: {}", task1(input.clone()));
    println!("Task 2: {}", task2(input.clone()));
}
